src/blocks/timewarrior.rs

```diff
@@ -0,0 +1,227 @@
+                ] {
+                    if let Some(value) = level {
+                        if (elapsed.num_minutes() as u64) >= *value {
+                            state = st;
+                            break;
+                        }
+                    }
+                }
+
+                values.insert("tags".into(), Value::text(tw.tags.join(" ")));
+
+                let elapsedstr =
+                    format!("{}:{:0>2}", elapsed.num_hours(), elapsed.num_minutes() % 60);
+                values.insert("elapsed".into(), Value::text(elapsedstr));
+
+                if let Some(annotation) = tw.annotation {
+                    values.insert("annotation".into(), Value::text(annotation));
+                }
+            }
+        }
+
+        widget.state = state;
+        widget.set_values(values);
+        api.set_widget(widget).await?;
+
+        select! {
+            _ = sleep(config.interval.0) => (),
+            event = api.event() => match event {
+                UpdateRequest => (),
+                Action(a) => {
+                    if a == "stop_continue" {
+                        stop_continue().await?;
+                    }
+                },
+            }
+        }
+    }
+}
+
+/// Raw output from timew
+#[derive(Deserialize, Debug)]
+struct TimewarriorRAW {
+    pub id: u32,
+    pub start: String,
+    pub tags: Vec<String>,
+    pub annotation: Option<String>,
+    pub end: Option<String>,
+}
+
+/// TimeWarrior entry
+#[derive(Debug, PartialEq)]
+struct TimewarriorData {
+    pub id: u32,
+    pub start: DateTime<chrono::offset::Utc>,
+    pub tags: Vec<String>,
+    pub annotation: Option<String>,
+    pub end: Option<DateTime<chrono::offset::Utc>>,
+}
+
+impl From<TimewarriorRAW> for TimewarriorData {
+    fn from(item: TimewarriorRAW) -> Self {
+        Self {
+            id: item.id,
+            tags: item.tags,
+            annotation: item.annotation,
+            start: DateTime::from_utc(
+                chrono::NaiveDateTime::parse_from_str(&item.start, "%Y%m%dT%H%M%SZ").unwrap(),
+                chrono::Utc,
+            ),
+            end: item.end.map(|v| {
+                DateTime::from_utc(
+                    chrono::NaiveDateTime::parse_from_str(&v, "%Y%m%dT%H%M%SZ").unwrap(),
+                    chrono::Utc,
+                )
+            }),
+        }
+    }
+}
+
+/// Format a DateTime given a format string
+#[allow(dead_code)]
+fn format_datetime(date: &DateTime<chrono::Utc>, format: &str) -> String {
+    date.format(format).to_string()
+}
+
+/// Execute "timew export now" and return the result
+async fn call_timewarrior() -> Result<String> {
+    let out = Command::new("timew")
+        .args(["export", "now"])
+        .output()
+        .await
+        .error("failed to run timewarrior")?
+        .stdout;
+    let output = std::str::from_utf8(&out)
+        .error("failed to read output from timewarrior (invalid UTF-8)")?
+        .trim();
+    Ok(output.to_string())
+}
```

Comment:
Oh, never mind, I didn't notice `.trim()`.

Reply:
Now that you mention it, I think it would work without the trim. I'm not sure if the JSON parser cares about spaces. I'll look into it

Reply:
For JSON, you don't even need to convert to a string :) https://docs.rs/serde_json/latest/serde_json/fn.from_slice.html

Reply:
Also I think it `process_timewarrior_data` can be moved inside this function, as it is only called with the result of `call_timewarrior()`. It would mean we would lose the tests, but the parsing seems simple enough anyway...